File: sdks/python/src/tidas_sdk/core/markdown.py
"""
Helpers for rendering entity data into Markdown.
"""
from __future__ import annotations

from typing import Any, Iterable

from ..core.multilang import MultiLangList


def _lang_and_text(entry: Any) -> tuple[str | None, str | None]:
    if entry is None:
        return None, None
    if isinstance(entry, dict):
        return entry.get("@xml:lang"), entry.get("#text")
    if hasattr(entry, "xml_lang") or hasattr(entry, "text"):
        return getattr(entry, "xml_lang", None), getattr(entry, "text", None)
    if isinstance(entry, str):
        return None, entry
    return None, None


def collect_texts(value: Any, lang: str = "en") -> list[str]:
    """
    Extract all text fragments for the requested language; fall back to any text.
    """
    entries: list[Any] = []
    if isinstance(value, MultiLangList):
        entries = list(value)
    elif isinstance(value, list):
        entries = value
    elif isinstance(value, dict):
        # Single entry shaped like MultiLang
        entries = [value]
    elif isinstance(value, str):
        return [value]
    else:
        return []

    lang_matches: list[str] = []
    fallback: list[str] = []
    for entry in entries:
        entry_lang, text = _lang_and_text(entry)
        if not text:
            continue
        if lang and entry_lang == lang:
            lang_matches.append(text)
        else:
            fallback.append(text)
    if lang_matches:
        return lang_matches
    return fallback


def pick_text(value: Any, lang: str = "en") -> str | None:
    """
    Return a single best-effort text fragment for the requested language.
    """
    if isinstance(value, MultiLangList):
        text = value.get_text(lang)
        return text or value.get_text(None)

    if isinstance(value, list):
        texts = collect_texts(value, lang)
        return texts[0] if texts else None

    if isinstance(value, dict):
        if "#text" in value:
            return value.get("#text")

    if isinstance(value, str):
        return value

    # Support other MultiLang-like helpers that expose get_text.
    getter = getattr(value, "get_text", None)
    if callable(getter):
        text = getter(lang)
        return text or getter(None)

    return None


def pick_short_description(ref: Any, lang: str = "en") -> str | None:
    """
    Extract the short description from a reference object or raw dict.
    """
    if ref is None:
        return None
    if isinstance(ref, list):
        for entry in ref:
            text = pick_short_description(entry, lang)
            if text:
                return text
        return None
    # Generated models expose `common_short_description`.
    attr = getattr(ref, "common_short_description", None)
    if attr is not None:
        text = pick_text(attr, lang)
        if text:
            return text
    if isinstance(ref, dict):
        if "common:shortDescription" in ref:
            text = pick_text(ref.get("common:shortDescription"), lang)
            if text:
                return text
        if "#text" in ref:
            return pick_text(ref, lang)
    return None


def join_texts(value: Any, lang: str = "en", *, sep: str = "\n\n") -> str | None:
    texts = collect_texts(value, lang)
    cleaned = [text.strip() for text in texts if text and text.strip()]
    return sep.join(cleaned) if cleaned else None


def format_number(value: Any) -> str:
    if value is None:
        return ""
    if isinstance(value, (int, float)):
        return f"{value:g}"
    try:
        return f"{float(str(value)) :g}"
    except Exception:
        return str(value)


def ensure_list(value: Any) -> list[Any]:
    if value is None:
        return []
    if isinstance(value, list):
        return value
    return [value]

File: sdks/typescript/src/utils/markdown.ts
import { MultiLangArray, MultiLangItemClass } from '../types/multi-lang-types';

type MultiLangLike =
  | MultiLangArray
  | MultiLangItemClass
  | { '@xml:lang'?: string; '#text'?: string }
  | Array<{ '@xml:lang'?: string; '#text'?: string }>;

function isMultiLangLike(value: any): value is MultiLangLike {
  if (!value) return false;
  if (value instanceof MultiLangArray || value instanceof MultiLangItemClass) return true;
  if (Array.isArray(value) && value.length && typeof value[0] === 'object') {
    return '#text' in value[0];
  }
  return typeof value === 'object' && '#text' in value;
}

export function ensureArray<T>(value: T | T[] | undefined | null): T[] {
  if (value === undefined || value === null) return [];
  return Array.isArray(value) ? value : [value];
}

export function collectTexts(value: any, lang = 'en'): string[] {
  if (!value) return [];
  if (typeof value === 'string') return [value];

  if (isMultiLangLike(value)) {
    const entries = Array.isArray(value) ? value : [value];
    const langMatches: string[] = [];
    const fallback: string[] = [];
    for (const entry of entries) {
      const text = (entry as any)['#text'];
      const entryLang = (entry as any)['@xml:lang'];
      if (!text) continue;
      if (lang && entryLang === lang) langMatches.push(String(text));
      else fallback.push(String(text));
    }
    return langMatches.length ? langMatches : fallback;
  }

  if (Array.isArray(value)) {
    return value.map(v => String(v));
  }
  if (typeof value === 'object' && '#text' in value) {
    return [String((value as any)['#text'])];
  }
  return [];
}

export function pickText(value: any, lang = 'en'): string | undefined {
  const texts = collectTexts(value, lang);
  return texts.length ? texts[0] : undefined;
}

export function joinTexts(value: any, lang = 'en', sep = '\n\n'): string | undefined {
  const texts = collectTexts(value, lang).map(item => item.trim()).filter(Boolean);
  return texts.length ? texts.join(sep) : undefined;
}

export function formatNumber(value: any): string {
  if (value === undefined || value === null) return '';
  if (typeof value === 'number') return Number.isFinite(value) ? value.toString() : String(value);
  const parsed = Number(value);
  return Number.isFinite(parsed) ? parsed.toString() : String(value);
}

export function pickShortDescription(ref: any, lang = 'en'): string | undefined {
  if (!ref) return undefined;
  if (Array.isArray(ref)) {
    for (const entry of ref) {
      const text = pickShortDescription(entry, lang);
      if (text) return text;
    }
    return undefined;
  }
  if (typeof ref === 'object') {
    if ('common:shortDescription' in ref) {
      return pickText((ref as any)['common:shortDescription'], lang);
    }
    if ('#text' in ref) return pickText(ref, lang);
  }
  return undefined;
}
